ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/BeaconStateMutators.java

```diff
@@ -98,18 +110,85 @@ public void initiateValidatorExit(MutableBeaconState state, int index) {
         .set(
             index,
             validator
-                .withExitEpoch(exitQueueEpoch)
+                .withExitEpoch(validatorExitContext.exitQueueEpoch)
                 .withWithdrawableEpoch(
-                    exitQueueEpoch.plus(specConfig.getMinValidatorWithdrawabilityDelay())));
+                    validatorExitContext.exitQueueEpoch.plus(
+                        specConfig.getMinValidatorWithdrawabilityDelay())));
+  }
+
+  /**
+   * This function implements an optimized version of exitQueueEpoch and exitQueueChurn calculation,
+   * compared to the `initiateValidatorExit` reference implementation.
+   *
+   * @param state
+   * @return
+   */
+  public Supplier<ValidatorExitContext> createValidatorExitContextSupplier(
+      final BeaconState state) {
+    return Suppliers.memoize(
```

Comment:
I don't clearly understand for how long is it cached. If it runs only once when client starts, it could be a risk for some reasons. Say, in new fork we will change `churnLimit`

Reply:
It can be reused only within the same state transition, so the slot (and `churnLimit`) is definitely not changing.

Reply:
ahh, got now, thank you for explanation.
A bit unclear anyway